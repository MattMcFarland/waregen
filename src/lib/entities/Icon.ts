import { X4Entity, X4EntityType } from "./X4Entity";
import { clone } from "lodash";

import {
  IconEntity,
  Attributes as IconAttributes
} from "@@/XMLTypes/X4LibraryIcons";

export type IconEntity = IconEntity;
export type IconAttributes = IconAttributes;

export const iconDefaults = (): IconAttributes =>
  clone({
    name: "",
    texture: "",
    height: 16,
    width: 16
  });

export class Icon extends X4Entity<IconEntity> {
  get name() {
    return this.xmlDef.Attributes.name;
  }
  set name(v: string) {
    if (typeof v !== "string") throw new TypeError(`${v} must be string`);
    if (typeof v === "string") this.xmlDef.Attributes.name = v;
  }
  get texture() {
    return this.xmlDef.Attributes.texture || iconDefaults().texture;
  }
  set texture(v: string | null | undefined) {
    if (typeof v !== "string") throw new TypeError(`${v} must be string`);
    if (typeof v === "string") this.xmlDef.Attributes.texture = v;
  }
  get height() {
    return this.xmlDef.Attributes.height || iconDefaults().height;
  }
  set height(v: number | null | undefined) {
    if (undefined == v) delete this.xmlDef.Attributes.height;
    if (typeof v !== "number") throw new TypeError(`${v} must be number`);
    if (typeof v === "number") this.xmlDef.Attributes.height = v;
  }
  get width() {
    return this.xmlDef.Attributes.width || iconDefaults().width;
  }
  set width(v: number | null | undefined) {
    if (undefined == v) delete this.xmlDef.Attributes.width;
    if (typeof v !== "number") throw new TypeError(`${v} must be number`);
    if (typeof v === "number") this.xmlDef.Attributes.width = v;
  }
  get active() {
    return this.xmlDef.Attributes.active || iconDefaults().active;
  }
  set active(v: string | null | undefined) {
    if (undefined == v) delete this.xmlDef.Attributes.active;
    if (typeof v !== "string") throw new TypeError(`${v} must be string`);
    if (typeof v === "string") this.xmlDef.Attributes.active = v;
  }
  get inactive() {
    return this.xmlDef.Attributes.inactive || iconDefaults().inactive;
  }
  set inactive(v: string | null | undefined) {
    if (undefined == v) delete this.xmlDef.Attributes.inactive;
    if (typeof v !== "string") throw new TypeError(`${v} must be string`);
    if (typeof v === "string") this.xmlDef.Attributes.inactive = v;
  }
  get selected() {
    return this.xmlDef.Attributes.selected || iconDefaults().selected;
  }
  set selected(v: string | null | undefined) {
    if (undefined == v) delete this.xmlDef.Attributes.selected;
    if (typeof v === "string") this.xmlDef.Attributes.selected = v;
  }
  constructor(
    constructOptions: IconAttributes = iconDefaults(),
    xmlDef?: IconEntity
  ) {
    super(
      X4EntityType.LIBRARY_ICON,
      "icon",
      iconDefaults(),
      xmlDef || {
        Attributes: constructOptions
      }
    );
  }
}

export default Icon;
