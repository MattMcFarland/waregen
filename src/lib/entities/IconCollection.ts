import { X4Entity, X4EntityType } from "./X4Entity";
import {
  IconEntity,
  Icons,
  Attributes as IconAttributes
} from "@@/XMLTypes/X4LibraryIcons";

import { Icon } from "./Icon";

export type IconEntity = IconEntity;
export type IconEntities = IconEntity[];

export interface IconIndex {
  [key: string]: Icon;
}
export class IconCollection extends X4Entity<Icons> {
  private iconIndex: IconIndex = {};
  protected onImport(_xmlDef: Icons) {
    delete this.iconIndex;
    this.iconIndex = {};
    delete this.xmlDef;
    this.xmlDef = { icon: [] };

    if (Array.isArray(_xmlDef.icon)) {
      _xmlDef.icon.forEach((iconEntity: IconEntity) => {
        const constructorOptions: IconAttributes = iconEntity.Attributes;
        const icon = new Icon(constructorOptions);
        this.add(icon);
      });
    }
  }
  add = (icon: Icon) => {
    if (this.iconIndex[icon.name]) {
      throw new Error(
        `icon ${icon.name} already exists, did you mean replaceWare?`
      );
    }
    this.iconIndex[icon.name] = icon;
    if (Array.isArray(this.xmlDef.icon)) {
      this.xmlDef.icon = [...this.xmlDef.icon, icon.__xmlDef];
    }
  };
  remove = (name: string) => {
    if (!this.iconIndex[name]) {
      throw new Error(`cant remove icon ${name} because it does not exist`);
    }
    delete this.iconIndex[name];
    if (Array.isArray(this.xmlDef.icon)) {
      this.xmlDef.icon = this.xmlDef.icon.filter(
        iconEntity => iconEntity.Attributes.name !== name
      );
    }
  };
  replace = (icon: Icon) => {
    this.remove(icon.name);
    this.add(icon);
  };
  extract = (wareId: string) => {
    return this.iconIndex[wareId];
  };
  constructor(optionalIconList?: Icon[]) {
    super(X4EntityType.LIBRARY_ICONS, "icons", {
      icon: []
    });
    if (Array.isArray(optionalIconList)) {
      optionalIconList.forEach((icon: Icon) => {
        this.add(icon);
      });
    }
  }
}
