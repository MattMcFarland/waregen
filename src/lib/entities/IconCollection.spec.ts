import { IconCollection } from "./IconCollection";
import { Icon, IconAttributes } from "./Icon";

const iconTest = (name: string): IconAttributes => ({
  name,
  texture: `assets/${name}`
});

describe("IconCollection", () => {
  describe("constructor", () => {
    const iconCollection = new IconCollection([
      new Icon(iconTest("A")),
      new Icon(iconTest("B")),
      new Icon(iconTest("C"))
    ]);
    test("success", () => {
      expect(iconCollection.extract("A")).toBeDefined();
      expect(iconCollection.extract("B")).toBeDefined();
      expect(iconCollection.extract("C")).toBeDefined();
    });
    test("instance.toXml()", () => {
      expect(iconCollection.toXml()).toMatchInlineSnapshot(`
"<icons>
  <icon name=\\"A\\" texture=\\"assets/A\\"/>
  <icon name=\\"B\\" texture=\\"assets/B\\"/>
  <icon name=\\"C\\" texture=\\"assets/C\\"/>
</icons>"
`);
    });
    test("instance.toJson()", () => {
      expect(iconCollection.toJson()).toMatchInlineSnapshot(
        `"{\\"icon\\":[{\\"Attributes\\":{\\"name\\":\\"A\\",\\"texture\\":\\"assets/A\\"}},{\\"Attributes\\":{\\"name\\":\\"B\\",\\"texture\\":\\"assets/B\\"}},{\\"Attributes\\":{\\"name\\":\\"C\\",\\"texture\\":\\"assets/C\\"}}]}"`
      );
    });
  });
  describe("instance.add()", () => {
    test("instance.toXml()", () => {
      const iconCollection = new IconCollection();
      iconCollection.add(new Icon(iconTest("A")));
      iconCollection.add(new Icon(iconTest("B")));
      expect(iconCollection.toXml()).toMatchInlineSnapshot(`
"<icons>
  <icon name=\\"A\\" texture=\\"assets/A\\"/>
  <icon name=\\"B\\" texture=\\"assets/B\\"/>
</icons>"
`);
    });
    test("instance.toJson()", () => {
      const iconCollection = new IconCollection();
      iconCollection.add(new Icon(iconTest("A")));
      iconCollection.add(new Icon(iconTest("B")));
      expect(iconCollection.toJson()).toMatchInlineSnapshot(
        `"{\\"icon\\":[{\\"Attributes\\":{\\"name\\":\\"A\\",\\"texture\\":\\"assets/A\\"}},{\\"Attributes\\":{\\"name\\":\\"B\\",\\"texture\\":\\"assets/B\\"}}]}"`
      );
    });
  });
  describe("instance.remove()", () => {
    test("instance.toXml()", () => {
      const iconCollection = new IconCollection();
      iconCollection.add(new Icon(iconTest("A")));
      iconCollection.add(new Icon(iconTest("B")));
      iconCollection.remove("A");
      expect(iconCollection.toXml()).toMatchInlineSnapshot(`
"<icons>
  <icon name=\\"B\\" texture=\\"assets/B\\"/>
</icons>"
`);
    });
    test("instance.toJson()", () => {
      const iconCollection = new IconCollection();
      iconCollection.add(new Icon(iconTest("A")));
      iconCollection.add(new Icon(iconTest("B")));
      iconCollection.remove("A");
      expect(iconCollection.toJson()).toMatchInlineSnapshot(
        `"{\\"icon\\":[{\\"Attributes\\":{\\"name\\":\\"B\\",\\"texture\\":\\"assets/B\\"}}]}"`
      );
    });
  });
  describe("instance.replace()", () => {
    test("instance.toXml()", () => {
      const iconCollection = new IconCollection();
      iconCollection.add(new Icon(iconTest("A")));
      const replacement = new Icon(iconTest("A"));
      replacement.texture = "REPLACED!";
      iconCollection.replace(replacement);
      expect(iconCollection.toXml()).toMatchInlineSnapshot(`
"<icons>
  <icon name=\\"A\\" texture=\\"REPLACED!\\"/>
</icons>"
`);
    });
    test("instance.toJson()", () => {
      const iconCollection = new IconCollection();
      iconCollection.add(new Icon(iconTest("A")));
      const replacement = new Icon(iconTest("A"));
      replacement.texture = "REPLACED!";
      iconCollection.replace(replacement);
      expect(iconCollection.toJson()).toMatchInlineSnapshot(
        `"{\\"icon\\":[{\\"Attributes\\":{\\"name\\":\\"A\\",\\"texture\\":\\"REPLACED!\\"}}]}"`
      );
    });
  });
  describe("instance.import()", () => {
    const testXml = () => {
      return `
      <icons>
      <icon name="missile_dumbfire_mk1" texture="assets/fx/gui/textures/PlayerHud/hud_ms_dumbfire_mk1.tga" height="32" width="64"></icon>
      <icon name="missile_dumbfire_mk2" texture="assets/fx/gui/textures/PlayerHud/hud_ms_dumbfire_mk2.tga" height="32" width="64"></icon>
      <icon name="missile_dumbfire_mk3" texture="assets/fx/gui/textures/PlayerHud/hud_ms_dumbfire_mk3.tga" height="32" width="64"></icon>
    
      <icon name="missile_guided_mk1" texture="assets/fx/gui/textures/PlayerHud/hud_ms_guided_mk1.tga" height="32" width="64"></icon>
      <icon name="missile_guided_mk2" texture="assets/fx/gui/textures/PlayerHud/hud_ms_guided_mk2.tga" height="32" width="64"></icon>
      <icon name="missile_guided_mk3" texture="assets/fx/gui/textures/PlayerHud/hud_ms_guided_mk3.tga" height="32" width="64"></icon>
    
      <icon name="missile_torpedo_mk1" texture="assets/fx/gui/textures/PlayerHud/hud_ms_torpedo_mk1.tga" height="32" width="64"></icon>
      <icon name="missile_torpedo_mk2" texture="assets/fx/gui/textures/PlayerHud/hud_ms_torpedo_mk2.tga" height="32" width="64"></icon>
      <icon name="missile_torpedo_mk3" texture="assets/fx/gui/textures/PlayerHud/hud_ms_torpedo_mk3.tga" height="32" width="64"></icon>
    
      <icon name="weapon_beam_mk1" texture="assets/fx/gui/textures/PlayerHud/hud_wp_beam_mk1.tga" height="32" width="64"></icon>
      <icon name="weapon_beam_mk2" texture="assets/fx/gui/textures/PlayerHud/hud_wp_beam_mk2.tga" height="32" width="64"></icon>
      <icon name="weapon_beam_mk3" texture="assets/fx/gui/textures/PlayerHud/hud_wp_beam_mk3.tga" height="32" width="64"></icon>            
      </icons>
      `;
    };
    test("toXML()", async () => {
      const testCollection = new IconCollection();
      await testCollection.import(testXml());
      expect(testCollection.toXml()).toMatchSnapshot();
    });
    test("with add", async () => {
      const testCollection = new IconCollection();
      await testCollection.import(testXml());
      testCollection.add(new Icon(iconTest("A")));
      expect(testCollection.toXml()).toMatchSnapshot();
    });
    test("with remove", async () => {
      const testCollection = new IconCollection();
      await testCollection.import(testXml());
      testCollection.remove("weapon_beam_mk1");
      testCollection.remove("weapon_beam_mk2");
      testCollection.remove("weapon_beam_mk3");
      expect(testCollection.toXml()).toMatchSnapshot();
    });
  });
});
