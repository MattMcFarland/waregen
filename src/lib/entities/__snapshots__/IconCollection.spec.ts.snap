// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconCollection instance.import() toXML() 1`] = `
"<icons>
  <icon name=\\"missile_dumbfire_mk1\\" texture=\\"assets/fx/gui/textures/PlayerHud/hud_ms_dumbfire_mk1.tga\\" height=\\"32\\" width=\\"64\\"/>
  <icon name=\\"missile_dumbfire_mk2\\" texture=\\"assets/fx/gui/textures/PlayerHud/hud_ms_dumbfire_mk2.tga\\" height=\\"32\\" width=\\"64\\"/>
  <icon name=\\"missile_dumbfire_mk3\\" texture=\\"assets/fx/gui/textures/PlayerHud/hud_ms_dumbfire_mk3.tga\\" height=\\"32\\" width=\\"64\\"/>
  <icon name=\\"missile_guided_mk1\\" texture=\\"assets/fx/gui/textures/PlayerHud/hud_ms_guided_mk1.tga\\" height=\\"32\\" width=\\"64\\"/>
  <icon name=\\"missile_guided_mk2\\" texture=\\"assets/fx/gui/textures/PlayerHud/hud_ms_guided_mk2.tga\\" height=\\"32\\" width=\\"64\\"/>
  <icon name=\\"missile_guided_mk3\\" texture=\\"assets/fx/gui/textures/PlayerHud/hud_ms_guided_mk3.tga\\" height=\\"32\\" width=\\"64\\"/>
  <icon name=\\"missile_torpedo_mk1\\" texture=\\"assets/fx/gui/textures/PlayerHud/hud_ms_torpedo_mk1.tga\\" height=\\"32\\" width=\\"64\\"/>
  <icon name=\\"missile_torpedo_mk2\\" texture=\\"assets/fx/gui/textures/PlayerHud/hud_ms_torpedo_mk2.tga\\" height=\\"32\\" width=\\"64\\"/>
  <icon name=\\"missile_torpedo_mk3\\" texture=\\"assets/fx/gui/textures/PlayerHud/hud_ms_torpedo_mk3.tga\\" height=\\"32\\" width=\\"64\\"/>
  <icon name=\\"weapon_beam_mk1\\" texture=\\"assets/fx/gui/textures/PlayerHud/hud_wp_beam_mk1.tga\\" height=\\"32\\" width=\\"64\\"/>
  <icon name=\\"weapon_beam_mk2\\" texture=\\"assets/fx/gui/textures/PlayerHud/hud_wp_beam_mk2.tga\\" height=\\"32\\" width=\\"64\\"/>
  <icon name=\\"weapon_beam_mk3\\" texture=\\"assets/fx/gui/textures/PlayerHud/hud_wp_beam_mk3.tga\\" height=\\"32\\" width=\\"64\\"/>
</icons>"
`;

exports[`IconCollection instance.import() with add 1`] = `
"<icons>
  <icon name=\\"missile_dumbfire_mk1\\" texture=\\"assets/fx/gui/textures/PlayerHud/hud_ms_dumbfire_mk1.tga\\" height=\\"32\\" width=\\"64\\"/>
  <icon name=\\"missile_dumbfire_mk2\\" texture=\\"assets/fx/gui/textures/PlayerHud/hud_ms_dumbfire_mk2.tga\\" height=\\"32\\" width=\\"64\\"/>
  <icon name=\\"missile_dumbfire_mk3\\" texture=\\"assets/fx/gui/textures/PlayerHud/hud_ms_dumbfire_mk3.tga\\" height=\\"32\\" width=\\"64\\"/>
  <icon name=\\"missile_guided_mk1\\" texture=\\"assets/fx/gui/textures/PlayerHud/hud_ms_guided_mk1.tga\\" height=\\"32\\" width=\\"64\\"/>
  <icon name=\\"missile_guided_mk2\\" texture=\\"assets/fx/gui/textures/PlayerHud/hud_ms_guided_mk2.tga\\" height=\\"32\\" width=\\"64\\"/>
  <icon name=\\"missile_guided_mk3\\" texture=\\"assets/fx/gui/textures/PlayerHud/hud_ms_guided_mk3.tga\\" height=\\"32\\" width=\\"64\\"/>
  <icon name=\\"missile_torpedo_mk1\\" texture=\\"assets/fx/gui/textures/PlayerHud/hud_ms_torpedo_mk1.tga\\" height=\\"32\\" width=\\"64\\"/>
  <icon name=\\"missile_torpedo_mk2\\" texture=\\"assets/fx/gui/textures/PlayerHud/hud_ms_torpedo_mk2.tga\\" height=\\"32\\" width=\\"64\\"/>
  <icon name=\\"missile_torpedo_mk3\\" texture=\\"assets/fx/gui/textures/PlayerHud/hud_ms_torpedo_mk3.tga\\" height=\\"32\\" width=\\"64\\"/>
  <icon name=\\"weapon_beam_mk1\\" texture=\\"assets/fx/gui/textures/PlayerHud/hud_wp_beam_mk1.tga\\" height=\\"32\\" width=\\"64\\"/>
  <icon name=\\"weapon_beam_mk2\\" texture=\\"assets/fx/gui/textures/PlayerHud/hud_wp_beam_mk2.tga\\" height=\\"32\\" width=\\"64\\"/>
  <icon name=\\"weapon_beam_mk3\\" texture=\\"assets/fx/gui/textures/PlayerHud/hud_wp_beam_mk3.tga\\" height=\\"32\\" width=\\"64\\"/>
  <icon name=\\"A\\" texture=\\"assets/A\\"/>
</icons>"
`;

exports[`IconCollection instance.import() with remove 1`] = `
"<icons>
  <icon name=\\"missile_dumbfire_mk1\\" texture=\\"assets/fx/gui/textures/PlayerHud/hud_ms_dumbfire_mk1.tga\\" height=\\"32\\" width=\\"64\\"/>
  <icon name=\\"missile_dumbfire_mk2\\" texture=\\"assets/fx/gui/textures/PlayerHud/hud_ms_dumbfire_mk2.tga\\" height=\\"32\\" width=\\"64\\"/>
  <icon name=\\"missile_dumbfire_mk3\\" texture=\\"assets/fx/gui/textures/PlayerHud/hud_ms_dumbfire_mk3.tga\\" height=\\"32\\" width=\\"64\\"/>
  <icon name=\\"missile_guided_mk1\\" texture=\\"assets/fx/gui/textures/PlayerHud/hud_ms_guided_mk1.tga\\" height=\\"32\\" width=\\"64\\"/>
  <icon name=\\"missile_guided_mk2\\" texture=\\"assets/fx/gui/textures/PlayerHud/hud_ms_guided_mk2.tga\\" height=\\"32\\" width=\\"64\\"/>
  <icon name=\\"missile_guided_mk3\\" texture=\\"assets/fx/gui/textures/PlayerHud/hud_ms_guided_mk3.tga\\" height=\\"32\\" width=\\"64\\"/>
  <icon name=\\"missile_torpedo_mk1\\" texture=\\"assets/fx/gui/textures/PlayerHud/hud_ms_torpedo_mk1.tga\\" height=\\"32\\" width=\\"64\\"/>
  <icon name=\\"missile_torpedo_mk2\\" texture=\\"assets/fx/gui/textures/PlayerHud/hud_ms_torpedo_mk2.tga\\" height=\\"32\\" width=\\"64\\"/>
  <icon name=\\"missile_torpedo_mk3\\" texture=\\"assets/fx/gui/textures/PlayerHud/hud_ms_torpedo_mk3.tga\\" height=\\"32\\" width=\\"64\\"/>
</icons>"
`;
